keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,418 @@
+such as the node startup scrips or an init container that inserts values from
+the Pod's Downward API.
+
+### Why not just keep using flags for instance-specific parameters?
+
+Flags have various problems, outlined in the [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c)
+doc.
+
+Leaving some parameters as flags while the rest are exposed via ComponentConfig
+results in an inconsistent API surface, and an inconsistent API versioning
+policy. Why fix only part of the problem when we can fix the whole thing?
+
+### Why not just auto-detect?
+
+One alternative that has been proposed in the past is to find a way to
+auto-detect values of instance-specific parameters and simply eliminate them
+from the configuration workflow. This is a great idea in theory, because it
+solves the problem by removing work for users, but it may be difficult in
+practice. For example, on a machine with multiple network cards, which IP should
+the kubelet use?
+
+Implementing the two-object approach does not prevent us from finding ways to
+auto-detect instance specific config in the future. It is worth noting that the
+_possibility_ of someday being able to auto-detect these has led to some
+hesitation to make a firm decision today. We have delayed for too long already,
+and a firm decision is needed to move forward.
+
+### Risks and Mitigations
+
+This creates an additional bucket for config parameters. We must be vigilant
+that only parameters that are truly instance-specific are added to the
+instance-specific Kind. We can provide warnings to developers via comments
+and tests to deter adding parameters to the instance-specific Kind that should
+instead be in the sharable Kind.
+
+For users deploying Pod templates, this solution can make the template more
+verbose. In the future, we can explore more elegant ways of generating files
+with instance-specific parameters correctly substituted. This proposal argues
+that the extra verbosity is worth the benefit of a consistent, versioned API for
+configuring core components.
+
+## Design Details
+
+ComponentConfig APIs are currently defined as a separate API group for each
+component, usually containing a single top-level Kind for configuring that
+component. The naming convention for the API group is `{component}.config.k8s.io`,
+and the convention for the Kind is `{Component}Configuration`. This KEP proposes
+that the naming convention for the new Kind be `{Component}InstanceConfiguration`.
+
+The canonical flag for providing a ComponentConfig file to a component is
+`--config`. This KEP proposes that the canonical flag for providing an
+instance-specific config file be `--instance-config`, and that the
+instance-specific object not initially be permitted as part of a yaml stream
+in the `--config` file (and vice-versa). This is for the sake of a simple
+implementation and can be enabled in the future, if we decide it is useful.
```

Comment:
With the last suggestion of multiple `--config` arguments, I miss a clear statement in this doc that do and do not want to tackle merging of parameters.

Reply:
Possible multiple `--config` flags (instead of `--instance-config` one) is probably the most flexible way of dealing with multiple kinds in such way. Some users may want to store the config in a couple of files (one instance specific and generated by an init container and another shared, provided by a config map volume source).
Others may desire to store both kinds in a single file and feed that in the component. That is already used by kubeadm (storing together InitConfiguration and ClusterConfiguration). kubeadm can also provision kubelet and static pod backed components (API Server, Scheduler and Controller Manager) in that way.

Implementation wise, we can probably add a single piece of code somewhere in component-base to deal with that in a common way.

Reply:
kubeadm has similar handling of different GVKs already. it only lacks the multiple `--config` flags - i.e. the part for loading multiple files.

https://github.com/kubernetes/kubernetes/blob/674695c544296c90e0a4fa0abdab71b3679a82c3/cmd/kubeadm/app/apis/kubeadm/types.go#L422

https://github.com/kubernetes/kubernetes/blob/e302aea142cb765e44c4842f3caebb4437db9dfb/cmd/kubeadm/app/util/marshal.go#L77

related PR by @obitech :
https://github.com/kubernetes/kubernetes/pulls/obitech

it can be done with a persistent `map[GroupVersionKind][]byte` during component startup.

- load the first `--config`
- documents in the file are processed with a non-strict unmarshal to gather the GVK and the `[]byte` is populated with the whole document.
- error out if multiple GroupKinds (GK) are provided in the same file and those GKs are not allowed to have multiple instances. [*]
- load another file from `--config`
- error out if a GK is present that was already loaded from a different file and this GK is not allowed to have multiple instances.
...
- repeat for all `--config` files.
...
- (string)unmarshall all `[]byte`s to public types / handle unknown and deprecated Versions.

[*] to set which can have multiple instances it would be sufficient to pass a `map[GroupKind]bool` to the utility function in component-base.

Reply:
Cool. Thanks for the pointers. That sounds pretty close to what we'd want.

I'm wondering if we can simplify that for unmarshaling inside the component, since the component only cares about its internal types (vs. kubeadm which has to deal with the external types)?

For example, loader would load and decode all files to internal types (and do the single-Kind check), and to extract you just pass the target object to an accessor. Something like this pseudocode:

```
NewFileLoader(Scheme, Codecs, map[GroupKind]bool) *FileLoader
(*FileLoader).LoadFiles([]string) (*ObjectSet, error) // load the files, decode into internal types (could be strict or non-strict depending on codecs), and return ObjectSet that has a type-match based accessor
(*ObjectSet).Get(runtime.Object) error // copy into runtime.Object (which should be an internal struct), or error if matching GroupKind can't be found in the set
```

Reply:
this is a much simple interface, indeed.

one question, would be if other components can end up with demand for unmarshaling external types the same way as kubeadm does.

potentially there could be support for both approaches.

Reply:
FWIW I like that this approach leaves it up to the user / wrapping tool entirely which fields are instance specific. 
Just supplying a common config in the first flag and let the second file containing arbitrarily many "instance specific" fields is pretty attractive.

+1 to considering this.